### PR TITLE
LSTM Realization

### PR DESCRIPTION
**Pull request by jdmattern-noaa**
Machine learning LSTM model and realization. Runs pre-trained LSTM models.

## Additions
Realization and model with params, config, and state. 

## Changes

Updates to CMake for optional compilation of the LSTM model due to LibTorch being required to run the LSTM.

## Testing

1. Unit tests added for LSTM Realization and Model

## Checklist

- [x] PR has an informative and human-readable title
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows project standards (link if applicable)
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

## Testing checklist (automated report can be put here)

1.

### Target Environment support

- [ ] Linux

### CODE REVIEW

File: doc/LSTM_MODEL.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  * Unit tests for the LSTM model and realization can then be built and run from the main directory with the following two commands:
+  
+      `cmake --build cmake-build-debug --target test_lstm`  <br />
+      
+      `./cmake-build-debug/test/test_lstm`  
+  
+  * The formulation config and required parameters to run the LSTM for a given catchment are below.
+
+## Formulation Config
+An example realization is ngen/data/lstm/example_lstm_realization_config.json.
```
Comment (hellkite500) [inline doc/LSTM_MODEL.md:26]: make this a link` [example_lstm_realization_config.json](ngen/data/lstm/example_lstm_realization_config.json)`

File: include/NGenConfig.h [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#ifndef NGEN_NGENCONFIG_H
```
Comment (hellkite500) [inline include/NGenConfig.h:1]: Isn't this a CMAKE generated file?  If so  it shouldn't be checked in.

File: include/realizations/catchment/Formulation_Constructors.hpp [hunk 3]
```diff
@@ -27,6 +31,10 @@ namespace realization {
         {"tshirt", create_formulation_constructor<Tshirt_Realization>()},
         {"tshirt_c", create_formulation_constructor<Tshirt_C_Realization>()},
         {"simple_lumped", create_formulation_constructor<Simple_Lumped_Model_Realization>()}
+#ifdef NGEN_LSTM_TORCH_LIB_ACTIVE
```
Comment (hellkite500) [inline include/realizations/catchment/Formulation_Constructors.hpp:34]: This is  the correct use of macro configuration,  but  I think we are going to want  to think a bit about  the long term impacts of using this feature.  @robertbartel  any thoughts from  you perspective on using optional libraries/realizations in this way?  My initial concern is in complicating  these  constructors/functions  for someone  trying to add  something  new.  Just something to think about for now.

File: include/realizations/catchment/LSTM_Realization.hpp [hunk 4]
```diff
@@ -0,0 +1,156 @@
+#ifndef NGEN_LSTM_REALIZATION_HPP
+#define NGEN_LSTM_REALIZATION_HPP
+
+#include "Catchment_Formulation.hpp"
+#include <unordered_map>
+
+#ifdef NGEN_LSTM_TORCH_LIB_ACTIVE
```
Comment (hellkite500) [inline include/realizations/catchment/LSTM_Realization.hpp:7]: Hmm, if LSTM.h  isn't  available, then this entire class cannot  be properly constructed.  Perhaps wrapping  this entire header with `ifdef`  is  more appropriate?

File: include/realizations/catchment/LSTM_Realization.hpp [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+            const std::vector<std::string>& get_required_parameters() override {
+                return REQUIRED_PARAMETERS;
+            }
+
+        private:
+            std::string catchment_id;
+            lstm::lstm_params params;
+            lstm::lstm_config config;
+           
+            #ifdef NGEN_LSTM_TORCH_LIB_ACTIVE
```
Comment (hellkite500) [inline include/realizations/catchment/LSTM_Realization.hpp:139]: See  previous comment.   If torch  isn't available, then this critical class member  cannot  be instantiated, so why expose any of the class?

File: include/utilities/CSV_Reader.h [hunk 6]
```diff
@@ -1,3 +1,6 @@
+#ifndef CSV_Reader_H
```
Comment (hellkite500) [inline include/utilities/CSV_Reader.h:1]: Good catch!

File: models/lstm/include/LSTM.h [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+        /** torch model */
+        torch::jit::script::Module model;
+
+        /** Parameter scaling */
+        ScaleParams scale;
+    };
+}
+
+#endif
```
Comment (hellkite500) [inline models/lstm/include/LSTM.h:141]: Add  a  comment after  this macro for what `if`  it is ending, i.e.  `#endif  //NGEN_LSTM_TORCH_LIB_ACTIVE`

File: models/lstm/include/lstm_state.h [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    };
+
+    inline void to_device( lstm_state& state, torch::Device device )
+    {
+      state.h_t.to(device);
+      state.c_t.to(device);
+    }
+} //namespace lstm
+
+#endif
```
Comment (hellkite500) [inline models/lstm/include/lstm_state.h:44]: add comment  for  which if  is  ending here

File: src/models/lstm/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+        ${PROJECT_SOURCE_DIR}/include
+        ${PROJECT_SOURCE_DIR}/include/utilities
+        ${PROJECT_SOURCE_DIR}/models/lstm/include
+        )
+
+target_link_libraries(models_lstm PUBLIC
+        Boost::boost                # Headers-only Boost
+        )
+
+if(LSTM_TORCH_LIB_ACTIVE)
```
Comment (hellkite500) [inline src/models/lstm/CMakeLists.txt:15]: If  Torch  isn't  active, is there any reason to do `add_library` or any of the other cmake  functions here?

File: src/models/lstm/LSTM.cpp [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+     * @param model_params Model parameters lstm_params struct.
+     */
+    lstm_model::lstm_model(lstm_config config, lstm_params model_params)
+            : config(config), model_params(model_params),
+            device( torch::Device(torch::kCPU) )
+    {
+        // ********** Set fluxes to null for now: it is bogus until first call of run function, which initializes it
+        fluxes = nullptr;
+        //manually set torch seed for reproducibility
+        torch::manual_seed(0);
```
Comment (hellkite500) [inline src/models/lstm/LSTM.cpp:58]: If we  are  only running trained networks,  I'm not sure the seed is important...

File: src/models/lstm/LSTM.cpp [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+        useGPU = config.useGPU && torch::cuda::is_available();
+        device = torch::Device( useGPU ? torch::kCUDA : torch::kCPU );
+
+        lstm_model::initialize_state(config.initial_state_path.c_str());
+ 
+        //FIXME need to initialize the model states by running a "warmup"
+        //so when the model is constructed, run it on N timesteps prior to prediction period
+        //OR read those states in and create the initial state tensors
+
+        std::cout<<"model_params.pytorch_model_path: " << config.pytorch_model_path;
```
Comment (hellkite500) [inline src/models/lstm/LSTM.cpp:68]: This  should get commented out until we have a better system for logging/verbosity

File: src/models/lstm/LSTM.cpp [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+
+        model = torch::jit::load(config.pytorch_model_path);
+        model.to( device );
+        // Set to `eval` model (just like Python)
+        model.eval();
+
+        //FIXME what is the SUPPOSED to do?
```
Comment (hellkite500) [inline src/models/lstm/LSTM.cpp:75]: Good question?

File: src/models/lstm/LSTM.cpp [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+
+        vector<double> h_vec;
+        vector<double> c_vec;
+        CSVReader reader(initial_state_path);
+        auto data = reader.getData();
+        std::vector<std::string> header = data[0];
+        //Advance the iterator to the first data row (skip the header)
+        auto row = data.begin();
+        std::advance(row, 1);
+        //Loop form first row to end of data
```
Comment (hellkite500) [inline src/models/lstm/LSTM.cpp:102]: Should probably  do a  quick sanity check that row has the required number of columns before the loop to avoid an out of bounds error on bad input:  `if( row.size() != 2 ){ throw error; }`

File: src/models/lstm/LSTM.cpp [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+      	// Run the model
+        auto output = model.forward(inputs).toTuple()->elements();
+      	//Get the outputs
+        double out_flow = lstm_model::denormalize( "obs", output[0].toTensor().item<double>() );
+        out_flow = out_flow * 0.028316847; //convert cfs to cms
+        fluxes = std::make_shared<lstm_fluxes>( lstm_fluxes( out_flow ) ) ;
+
+        current_state = std::make_shared<lstm_state>( lstm_state(output[1].toTensor(), output[2].toTensor()) );
+
+        return 0;
```
Comment (hellkite500) [inline src/models/lstm/LSTM.cpp:188]: Need to document the  return meaning of this function (i.e. 0 mean run successful, something else  is error)

**Review by hellkite500 — CHANGES_REQUESTED** [review]
A couple small changes (mostly comments)  are in order, but otherwise, looks  pretty  good!

**Comment by hellkite500** [comment]
Also, if #212 gets merged, then the example realization configs will break and will need to be updated with correct idenenities.

File: include/realizations/catchment/LSTM_Realization.hpp (hunk 5 above)
Comment (jdmattern-noaa) [reply include/realizations/catchment/LSTM_Realization.hpp:139]: Good points. Wrapping entire class now.

File: src/models/lstm/LSTM.cpp (hunk 10 above)
Comment (jdmattern-noaa) [reply src/models/lstm/LSTM.cpp:58]: Removing.

File: src/models/lstm/CMakeLists.txt (hunk 9 above)
Comment (jdmattern-noaa) [reply src/models/lstm/CMakeLists.txt:15]: Good point. I changed it all to go within the if.

File: src/models/lstm/LSTM.cpp (hunk 12 above)
Comment (jmframe) [reply src/models/lstm/LSTM.cpp:73]: no_grad Disables gradient calculations on the tensors, since we don't need them in the forward pass. This reduces memory consumption.

File: models/lstm/include/LSTM.h [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+#define LSTM_H
+
+#include "all.h"
+#include "lstm_fluxes.h"
+#include "lstm_params.h"
+#include "lstm_config.h"
+#include "lstm_state.h"
+#include <unordered_map>
+
+#ifdef NGEN_LSTM_TORCH_LIB_ACTIVE
```
Comment (hellkite500) [inline models/lstm/include/LSTM.h:11]: Can we move this macro to the top of the file?  again, no need to include headers even if we don't have torch?

File: src/models/lstm/LSTM.cpp (hunk 14 above)
Comment (hellkite500) [reply src/models/lstm/LSTM.cpp:196]: Add this to the doc string for the function under `@return`

File: test/CMakeLists.txt [hunk 16]
```diff
@@ -79,6 +83,13 @@ add_automated_test(
         core/catchment/giuh/GIUH_Test.cpp
 )
 
+#target_link_libraries
```
Comment (hellkite500) [inline test/CMakeLists.txt:86]: probably need to wrap this with `if(LSTM_TORCH_LIB_ACTIVE)` as well

**Review by hellkite500 — COMMENTED** [review]
Couple minor edits, but overall in good shape pending a rebase and ensuring that the hydrofabric refactor that got merged doesn't break anything.

**Comment by hellkite500** [comment]
Rebase, retest, and address last few comments then merge

**Comment by jdmattern-noaa** [comment]
@hellkite500 Rebased on top of master, made updates from your above feedback, unit tests pass, and updated the lstm example config and doc with the updates to the hydrofabric and that runs successfully.

**Review by hellkite500 — APPROVED** [review]

File: doc/LSTM_MODEL.md [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+  * Unit tests for the LSTM model and realization can then be built and run from the main directory with the following two commands:
+  
+      `cmake --build cmake-build-debug --target test_lstm`  <br />
+      
+      `./cmake-build-debug/test/test_lstm`  
+  
+  * The formulation config and required parameters to run the LSTM for a given catchment are below.
+
+## Formulation Config
+An example realization is https://github.com/NOAA-OWP/ngen/data/lstm/example_lstm_realization_config.json.
```
[the remark this answers is not in the document]
Comment (hellkite500) [reply doc/LSTM_MODEL.md]: I was actually implying an actual hyperref, but good catch.

`An [example realization](https://github.com/NOAA-OWP/ngen/data/lstm/example_lstm_realization_config.json)`.

**Review by hellkite500 — APPROVED** [review]
